### PR TITLE
Add odsp driver support to upload container type summary

### LINKED ISSUE #8079: ODSP support container upload in addition to channel upload when unploading protocol tree

### PR DESCRIPTION
**Pull request by jatgarg**
Fixes: https://github.com/microsoft/FluidFramework/issues/8079
This adds support for odsp driver to upload container type summary which includes the ".protocol" summary too along with app summary.
Currently there is no support at server to handle the full summary, so we get the SummaryNack in case we do upload the full summary in odsp driver.
But this is behind feature gate **"SummarizeProtocolTree"** in loader layer, so I think we don't need to add another feature gate for driver as if loader is sending the protocol tree then in that case odsp driver does have to upload everything, so I think another feature gate might not make sense. Let me know your thoughts @anthony-murphy .

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts [hunk 1]
```diff
@@ -55,6 +55,7 @@ export class OdspSummaryUploadManager {
                 lastSummaryProposalHandle: this.lastSummaryProposalHandle,
             });
         }
+        assert(context.ackHandle !== undefined, "Parent summary should always exist");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts:58]: Please see that (not yet shipping) code:
```ts
                    await this.storageService.uploadSummaryWithContext(summary, {
                        referenceSequenceNumber: 0,
                        ackHandle: undefined,
                        proposalHandle: undefined,
                    });
```

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts [hunk 2]
```diff
@@ -69,19 +69,19 @@ export class OdspSummaryUploadManager {
...
-            "",
         );
         const snapshot: IOdspSummaryPayload = {
             entries: snapshotTree.entries!,
             message: "app",
             sequenceNumber: referenceSequenceNumber,
             // no ack handle implies this is initial summary after empty file creation.
             // send container payload so server will use it without a summary op
-            type: parentHandle === undefined ? "container" : "channel",
+            type: containsProtocolTree || parentHandle === undefined ? "container" : "channel",
```
Comment (anthony-murphy) [inline packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts:84]: i'd like a sperate feature gate here. as we do run with summarizeProtocolTree in stress and it works as a channel upload, and i'd like to keep that working

**Review by anthony-murphy — APPROVED** [review]

**Comment by GaryWilber** [comment]
Regarding `Currently there is no support at server to handle the full summary, so we get the SummaryNack in case we do upload the full summary in odsp driver.`
I will need to make a change on my side to support this. Is it possible to have the client include a new flag in the summary op to indicate that it uploaded the container tree? There is an `includesProtocolTree` [flag set today ](https://github.com/microsoft/FluidFramework/blob/3ea8397949074f28f4fe2929de06859130d8ddcc/packages/loader/container-loader/src/container.ts#L1629) - maybe another flag can be added, such as `isContainerTree`

**Comment by GaryWilber** [comment]
> supporting a big matrix of unshipped features is not great. i'm fine if we just switch the server behavior. just let use know when you assume includesProtocolTree means container upload and we'll make the switch too. maybe do a server side flag, and we'll just coordinate rollout as none of this is in prod.

The issue I'm facing is that the server needs to know what kind of tree the client uploaded. 

There are a total of 3 cases:
1. Client uploads a channel tree without a .protocol tree in it 
2. Client uploads a channel tree with a .protocol tree
3. Client uploads a container tree with a .protocol tree

The server has to run different logic depending on each case:
1. Server commits the summary by POSTing a container, which is a server generated .protocol tree + the client created .app tree
2. Server fetches the client created .protocol tree, verifies the .protocol tree is valid, and POSTs that channel tree as a container.
3. Server _should_ not do anything (no calls to ODSP) because the client already did the commit.

The server submits a summaryAck after processing the summary in each case.

The server knows case 1 and 2 are different because of the `includesProtocolTree` flag.

However the server cannot determine the difference between case 2 and 3 because the flag is set for both. This means the server will have to figure out if the client uploaded a channel or container tree by calling ODSP. So "case 3" will actually involve a call to ODSP now, which is working against the purpose of what we want - I don't want Push to have to call ODSP when the client already committed the summary.

**Comment by jatgarg** [comment]
> > supporting a big matrix of unshipped features is not great. i'm fine if we just switch the server behavior. just let use know when you assume includesProtocolTree means container upload and we'll make the switch too. maybe do a server side flag, and we'll just coordinate rollout as none of this is in prod.
> 
> The issue I'm facing is that the server needs to know what kind of tree the client uploaded.
> 
> There are a total of 3 cases:
> 
> 1. Client uploads a channel tree without a .protocol tree in it
> 2. Client uploads a channel tree with a .protocol tree
> 3. Client uploads a container tree with a .protocol tree
> 
> The server has to run different logic depending on each case:
> 
> 1. Server commits the summary by POSTing a container, which is a server generated .protocol tree + the client created .app tree
> 2. Server fetches the client created .protocol tree, verifies the .protocol tree is valid, and POSTs that channel tree as a container.
> 3. Server _should_ not do anything (no calls to ODSP) because the client already did the commit.
> 
> The server submits a summaryAck after processing the summary in each case.
> 
> The server knows case 1 and 2 are different because of the `includesProtocolTree` flag.
> 
> However the server cannot determine the difference between case 2 and 3 because the flag is set for both. This means the server will have to figure out if the client uploaded a channel or container tree by calling ODSP. So "case 3" will actually involve a call to ODSP now, which is working against the purpose of what we want - I don't want Push to have to call ODSP when the client already committed the summary.

Yes, I think that is what Tony is saying. For now, even for 3rd you can do same as you do in 2 because it will not be worse than the current state. Then once we are confident, w can enable the flag permanently and then you can stop checking the type and eliminate 2nd state.

**Pull request merged**